exporter/newrelicexporter/transformer.go

```diff
@@ -0,0 +1,370 @@
+	}
+
+	if metric.MetricDescriptor.Description != "" {
+		attrs[descriptionAttrKey] = metric.MetricDescriptor.Description
+	}
+
+	attrs[collectorNameKey] = name
+	attrs[collectorVersionKey] = version
+	attrs[serviceNameKey] = t.ServiceName
+
+	return attrs
+}
+
+func (t *transformer) MergeAttributes(base map[string]interface{}, lk []*metricspb.LabelKey, lv []*metricspb.LabelValue) (map[string]interface{}, error) {
+	if len(lk) != len(lv) {
+		return nil, fmt.Errorf("number of label keys (%d) different than label values (%d)", len(lk), len(lv))
+	}
+
+	attrs := make(map[string]interface{}, len(base)+len(lk))
+
+	for k, v := range base {
+		attrs[k] = v
+	}
+	for i, k := range lk {
+		v := lv[i]
+		if v.Value != "" || v.HasValue {
+			attrs[k.Key] = v.Value
+		}
+	}
+	return attrs, nil
+}
+
+func (t *transformer) Gauge(name string, attrs map[string]interface{}, point *metricspb.Point) telemetry.Metric {
+	now := time.Now()
+	if point.Timestamp != nil {
+		now = t.Timestamp(point.Timestamp)
+	}
+
+	m := telemetry.Gauge{
+		Name:       name,
+		Attributes: attrs,
+		Timestamp:  now,
+	}
+
+	switch t := point.Value.(type) {
+	case *metricspb.Point_Int64Value:
+		m.Value = float64(t.Int64Value)
+	case *metricspb.Point_DoubleValue:
+		m.Value = t.DoubleValue
+	}
+
+	return m
+}
+
+func (t *transformer) DeltaSummary(name string, attrs map[string]interface{}, start time.Time, point *metricspb.Point) telemetry.Metric {
+	now := time.Now()
+	if point.Timestamp != nil {
+		now = t.Timestamp(point.Timestamp)
+	}
+
+	m := telemetry.Summary{
+		Name:       name,
+		Attributes: attrs,
+		Timestamp:  start,
+		Interval:   now.Sub(start),
+	}
+
+	switch t := point.Value.(type) {
+	case *metricspb.Point_DistributionValue:
+		m.Count = float64(t.DistributionValue.Count)
+		m.Sum = t.DistributionValue.Sum
+	case *metricspb.Point_SummaryValue:
+		m.Count = float64(t.SummaryValue.Count.Value)
+		m.Sum = t.SummaryValue.Sum.Value
+	}
+
+	return m
+}
+
+func (t *transformer) CumulativeCount(name string, attrs map[string]interface{}, start time.Time, point *metricspb.Point) telemetry.Metric {
+	var value float64
+	switch t := point.Value.(type) {
+	case *metricspb.Point_Int64Value:
+		value = float64(t.Int64Value)
+	case *metricspb.Point_DoubleValue:
+		value = t.DoubleValue
+	}
+
+	now := time.Now()
+	if point.Timestamp != nil {
+		now = t.Timestamp(point.Timestamp)
+	}
+
+	count, valid := t.DeltaCalculator.CountMetric(name, attrs, value, now)
+
+	// This is the first measurement or a reset happened.
+	if !valid {
+		count = telemetry.Count{
+			Name:       name,
+			Attributes: attrs,
+			Value:      value,
+			Timestamp:  start,
+			Interval:   now.Sub(start),
+		}
+	}
+
+	return count
+}
+
+func (t *transformer) CumulativeSummary(name string, attrs map[string]interface{}, start time.Time, point *metricspb.Point) telemetry.Metric {
+	var sum, count float64
+	switch t := point.Value.(type) {
+	case *metricspb.Point_DistributionValue:
```

Comment:
Since this is an unexported struct we own the caller as well as this receiver. I therefore didn't check for types not `Point_DistributionValue` given it won't be called with anything other than this.

If it is called with a different type the sum/count of the exported Summary will just be `0`.

I could add a default panic here with some tests to ensure the panic is never called if you think it is worth the effort.

Reply:
good point that we have control over the call sites of the method. i could go either way with adding a more explicit failure mode; it would be helpful if/when we come back to expand on the exported metric types, for example. i'm not sure how likely that situation will be in the future, so i'll defer to your judgement on the value of adding that. thanks for the explanation regardless!